lib/Parse/ParsePattern.cpp

```diff
@@ -1292,12 +1292,19 @@ ParserResult<Pattern> Parser::parseMatchingPattern(bool isExprBasic) {
   if (Tok.isAny(tok::kw_var, tok::kw_let) ||
       (Context.LangOpts.hasFeature(Feature::ReferenceBindings) &&
        Tok.isAny(tok::kw_inout))) {
+    if (inIsCaseExpr) {
+      auto newBindingState = PatternBindingState(Tok);
+      unsigned index = *newBindingState.getSelectIndexForIntroducer();
+      diagnose(Tok, diag::var_pattern_in_is_case_expr, index);
+      return nullptr;
+    }
+
```

Comment:
Bindings can also be nested in expressions, which may become patterns during pattern resolution. I think it would be best to move this diagnostic into the `ResolvePattern` walker